internal/expr/scan.go

```diff
@@ -0,0 +1,196 @@
+package expr
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+)
+
+func (re *ResultExpr) One(args ...any) error {
+	if ok, err := re.Next(); err != nil {
+		return err
+	} else if !ok {
+		return fmt.Errorf("cannot return one row: no results")
+	}
+	err := re.Decode(args...)
+	if err != nil {
+		return err
+	}
+	re.Close()
+	return nil
+}
+
+// getTypes returns the types in the order they appear in the query
+func getTypes(outs []field) []reflect.Type {
+	isDup := make(map[reflect.Type]bool)
+	ts := []reflect.Type{}
+	for _, out := range outs {
+		if t := out.structType; !isDup[t] {
+			isDup[t] = true
+			ts = append(ts, t)
+		}
+	}
+	return ts
+}
+
+// All returns a slice containing all rows returned in the query. Each row is
+// a slice of any types that contains all structs mentioned in the output expressions
+// of the query.
+func (re *ResultExpr) All() ([][]any, error) {
+	var s [][]any
+
+	ts := getTypes(re.outputs)
+
+	for {
+		if ok, err := re.Next(); err != nil {
+			return [][]any{}, err
+		} else if !ok {
+			break
+		}
+
+		rs := []any{}
+		for _, t := range ts {
+			rp := reflect.New(t)
+			// We need to unwrap the struct inside the interface{}.
+			r := rp.Elem()
+			err := re.decodeValue(r)
```

Comment:
It's slightly more efficient this way because we dont do all the checks of Decode, also `reflect.New` returns a `reflect.Value`. We would also have to get the `Interface` and make it a first class value only for it to be reflected on again in Decode which seems like a lot of wasted expensive reflection.